README.md
# Agilizer Source for Jira

[![Go Report Card](https://goreportcard.com/badge/github.com/rchampourlier/agilizer-source-jira)](https://goreportcard.com/report/github.com/rchampourlier/agilizer-source-jira)
[![Build Status](https://travis-ci.com/rchampourlier/agilizer-source-jira.svg?branch=master)](https://travis-ci.com/rchampourlier/agilizer-source-jira)
[![Coverage Status](https://coveralls.io/repos/github/rchampourlier/agilizer-source-jira/badge.svg)](https://coveralls.io/github/rchampourlier/agilizer-source-jira)

## Objective

The goal of this project is to enable teams using Jira to perform analyses on Jira projects through a SQL database, enabling the use of user-friendly tools like Metabase, more powerful ones like Superset or even advanced data analysis toolsets (Python, R...).

## Use cases

Some examples of easy requests that can be done using Metabase.

### Identify issues with many comments

![](doc/metabase-example-issues-with-many-comments.png)

### Track number of bugs created over time

![](doc/metabase-example-created-bugs-per-priority-over-time.png)

## How it works

The tool will connect to Jira using the API and fetch all issues. For each issue:

- a simplified representation of the issue is stored in the `jira_issues_states` table,
- a set of events is created in the `jira_issues_events` to represent the updates that occurred on the issue (e.g. `created`, `comment_added`, `status_changed`).

The tool will perform a request to only retrieve the issues modified since the last synchronization, using the timestamp of the last event. All corresponding issues will be processed to generate new events as needed.

### Requirements

- A PostgreSQL database
- Jira username and password
- `cp .env.example` updated as necessary

### How to use

#### 0. Clone the repo

```
go get github.com/rchampourlier/agilizer-source-jira
```

NB: you should follow Go conventions (e.g. `GOPATH`, dependencies management, etc.) or at least have your environment working and know it well enough.

#### 1. Create your `.env` file

```
cp .env.example .env
```

Now, edit the file and set the following values:

- `JIRA_USERNAME`: a valid Jira username (you should probably have a dedicated user for this and not use your personal user)
- `JIRA_PASSWORD`: the corresponding password
- `DB_URL`: the URL to the database where you want to push Jira records to (it should look like this: `postgres://USER:PASSWORD@HOST:5432/DB_NAME`)

NB: if you face Postgres SSL-related issues, try adding `?sslmode=disable` at the end of your `DB_URL`.

#### 2. DB initialization and initial synchronization

```
source .env
go run *.go reset
```

#### 3. Incremental synchronization

```
source .env
go run *.go sync
```

_NB: the DB must have been initialized and a first synchronization done._

### How to contribute / customize

#### Run tests

```
make test
```

#### How to change the generated state and event records

##### Add a new field to the _Jira Issue States_

- **In `store/pgstore.go`**
  - In `CreateTables(..)`, add the column for the new field to the `jira_issues_states` table.
  - In `insertIssueState(..)`, add the new value in the `INSERT`.
- **In `store/store.go`**
  - Change the `IssueState struct` to add the new field.
- **[Optional] If you want to add the field to the tests (necessary if the field is mandatory or you do some operation - e.g. mapping or conversion), in `store/mockstore.go`**
  - Update `ReplaceIssueStateAndEvents(..)` to check the value for the new field.
- **In `jira/mapping/mapper.go`**
  - Change `IssueStateFromIssue(..)` to generate the correct `store.IssueState` for your issue, adding the new field. (This is where you will do the mapping with custom fields.)
- Run the tests and fix/update as necessary.

NB: you can use the `explore-custom-fields` action on the command line to get custom fields mappings.

##### Generate new kinds of _Jira Issue Events_

For now, the following events are generated from the issue's data:

- `created`
- `comment_added`
- `status_changed`
- `assignee_changed`

If you want to add new kinds of events:

- **In `jira/mapping/mapper.go`**
  - Edit `IssueEventsFromIssue(..)` to generate your new events for each issue processed. You can see how the existing events are generated.
  - Update the corresponding tests in `jira/mapping/mapper_test.go`
- [Optional] If you need to change the _Jira Issue Events_ structure to add columns related to your new events, you can follow the instructions for _Jira Issue States_ above, there is not much difference (unless you should look for event-related functions!).
- As always, do not forget to run tests and fix/update them if necessary.

## Troubleshooting

### SSL issues with Postgres

Add `?sslmode=disable` at the end of your DB URL.

## Contribution

The project is not opened to contribution at the moment.

## License

MIT



HISTORY.md
# History

## 2018-08-08

- Adding `assignee_changed` events
- Minor bug fix (use `reset` instead of `init` now)
- Added documentation to fix missing information

## 2018-08-03

- Deployed on JobTeaser's Kubernetes development cluster. Took me less than 2 hours, from scratch. With the help of Gearnode :) Yey!

## 2018-07-29

- Enabled deployment to Heroku. Added `dep` as dependency manager. Abandoned deployment to Heroku because it doesn't seem usable from the Scheduler plugin that is needed to schedule task executions.
- Kept the Heroku's build stack to generate the Docker image for now. Will use this to try deploying to AWS ECS or Fargate. Achieved deploying the container image to AWS ECR, but going further is a pain in the $#!. Seems like scheduling will not be easy either, and Fargate, well... could not really find where Fargate was.
- Trying GCP AppEngine now. I will finally switch to Serverless if I can't have this easy setup to work without deploying my own Docker server. Well, after some time trying, it may not be simpler here. Abandoning deployment for now, will use our Kubernetes cluster instead, since it may be the easiest solution.
- Going back to AWS. One more time. But still failing, not being capable of overriding the container's command to pass the `sync` argument :(

## 2018-07-28

- Finished with unit tests.
- Added configuration to the repository to run tests using TravisCI and check coverage with Coveralls.
- Repository configured on Github to display badges and do some CI on pull request changes.
- Adding some documentation to help with usage and changes.

## 2018-07-17

Continuing to implement unit tests. Facing an issue for `sync_test.go`, because the tests seem to stop executing before the goroutines have all been executed.

- Created mocks for new interfaces `jira.Client` and `store.Store`
- Extracted shared code for store to `store/store.go`
- Renamed `store.Store` to `store.PGStore` to clarify its a Postgres backend

## 2018-07-15

**Adding unit tests**

Also performed some modifications to the structure of the code to make it more testable. For example:

- Refactored `store` to perform the deletion of previous records and insertion of new records for a given issue in a single operation, wrapped in a transaction. This simplifies the tests while making the operation more robust.
- Extracted `jira.Client` to a specific `jira/client` package, adding an interface in `jira/client.go` to enable the addition of a `MockClient` to test the synchronization code easily.

## 2018-07-14

- [x] Handle incremental synchronization to enable the synchronization to be performed regularly (e.g. every 10 minutes).
  - [x] Fetch only issues updated after max `issue_updated_at`
  - [x] Drop previous states and events for the updated issues in an incremental sync
- [ ] Store a local cache copy of issue to enable faster reprocessing when changing only the mapping --> WON'T DO
  Abandoned because storing the raw issue as JSON is not reversible (unmarshaling back to a `jira.Issue` fails, because the dates are not marshaled), so it would be too much work to cache the ra
w issue (it would require either to fetch directly from the API, without using `go-jira`, or changing the marshaling).
- [ ] Synchronize transparently with a DB table switch --> WON'T DO
  Doesn't seem necessary with the incremental sync, for now at least.

## Earlier

### Data model

- [x] For `jira_issue_events` replace `..._author` columns by a single `event_author` one
- [x] Add appropriate columns to `jira_issues_events` to simplify analysis (e.g. tribe, project, issue_type...)

### Bugs

- [x] Fix synchronization bug

Identified issue causing the bug: `DI-146`

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x12511bf]

goroutine 21 [running]:
main.issueEventsFromIssue(0xc421645080, 0xc420656346, 0x6, 0xc421645080)
        /Users/rchampourlier/dev/agilizer-source-jira/jiramapping.go:51 +0xbf
main.(*jiraClient).performSync.func1(0x1285200, 0xc42071c630, 0x0, 0x0)
        /Users/rchampourlier/dev/agilizer-source-jira/sync.go:29 +0xd6
github.com/Jeffail/tunny.(*closureWorker).Process(0xc42000e140, 0x1285200, 0xc42071c630, 0x0, 0x0)
        /Users/rchampourlier/go/src/github.com/Jeffail/tunny/tunny.go:73 +0x3d
github.com/Jeffail/tunny.(*workerWrapper).run(0xc420186660)
        /Users/rchampourlier/go/src/github.com/Jeffail/tunny/worker.go:101 +0x341
created by github.com/Jeffail/tunny.newWorkerWrapper
        /Users/rchampourlier/go/src/github.com/Jeffail/tunny/worker.go:70 +0x141
panic: worker was closed

goroutine 1346 [running]:
github.com/Jeffail/tunny.(*Pool).Process(0xc420024ac0, 0x1285200, 0xc42071c630, 0x0, 0x0)
        /Users/rchampourlier/go/src/github.com/Jeffail/tunny/tunny.go:163 +0x13b
created by main.(*jiraClient).performSync.func2
        /Users/rchampourlier/dev/agilizer-source-jira/sync.go:40 +0xcb
```

- [x] Fix DB bug (it seems it was an issue with the number of connections to the DB)

```
2018/06/17 18:50:50 error in `insertIssueEvent`: pq: SSL is not enabled on the server
exit status 1
```

### Inserting issues (state)

- [x] Fix resolutiondate parsing
- [x] Insert issues in the DB (state, not events), just like the existing source.
- [x] Add custom fields to issue states

### Misc

- [x] Fix bug, pool closed once the search is complete, without waiting the fetches to finish
- [X] Convert event's `time` to a datetime column

### Connection with AgilizerSource PostgreSQL db

- [x] DB connection
- [x] DB initialization (create tables...)
- [x] Write events to DB instead of logging



db/db.go
package db

import (
	"database/sql"
	"fmt"
	"log"
	"os"
	"time"

	_ "github.com/lib/pq"
)

type DB struct {
	*sql.DB
}

func NewDB() *DB {
	return &DB{openDB()}
}

func (db *DB) Close() error {
	return db.DB.Close()
}

// MaxOpenConns defines the maximum number of open connections
// to the DB.
const MaxOpenConns = 5 // for Heroku Postgres

type IssueEvent struct {
	EventTime        time.Time
	EventKind        string
	EventAuthor      string
	IssueKey         string
	CommentBody      *string
	StatusChangeFrom *string
	StatusChangeTo   *string
}

type IssueState struct {
	CreatedAt         time.Time
	UpdatedAt         time.Time
	Key               string
	Project           *string
	Status            *string
	ResolvedAt        *time.Time
	Priority          *string
	Summary           *string
	Description       *string
	Type              *string
	Labels            *string
	Reporter          *string
	Assignee          *string
	DeveloperBackend  *string
	DeveloperFrontend *string
	Reviewer          *string
	ProductOwner      *string
	BugCause          *string
	Epic              *string
	Tribe             *string
	Components        *string
	FixVersions       *string
}

func (db *DB) InsertIssueEvent(e IssueEvent, s IssueState) {
	query := `
	INSERT INTO jira_issues_events (
		event_time,
		event_kind,
		event_author,
		comment_body,
		status_change_from,
		status_change_to,
		issue_key,
		issue_created_at,
		issue_updated_at,
		issue_project,
		issue_status,
		issue_resolved_at,
		issue_priority,
		issue_summary,
		issue_description,
		issue_type,
		issue_labels,
		issue_assignee,
		issue_developer_backend,
		issue_developer_frontend,
		issue_reviewer,
		issue_product_owner,
		issue_bug_cause,
		issue_epic,
		issue_tribe,
		issue_components,
		issue_fix_versions
	)
	VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26, $27);
	`

	rows, err := db.Query(
		query,
		e.EventTime,
		e.EventKind,
		e.EventAuthor,
		e.CommentBody,
		e.StatusChangeFrom,
		e.StatusChangeTo,
		e.IssueKey,
		s.CreatedAt,
		s.UpdatedAt,
		s.Project,
		s.Status,
		s.ResolvedAt,
		s.Priority,
		s.Summary,
		s.Description,
		s.Type,
		s.Labels,
		s.Assignee,
		s.DeveloperBackend,
		s.DeveloperFrontend,
		s.Reviewer,
		s.ProductOwner,
		s.BugCause,
		s.Epic,
		s.Tribe,
		s.Components,
		s.FixVersions,
	)
	if err != nil {
		log.Fatalln(fmt.Errorf("error in `InsertIssueEvent`: %s", err))
	}
	rows.Close()
}

func (db *DB) InsertIssueState(s IssueState) {
	query := `
	INSERT INTO jira_issues_states (
		issue_created_at,
		issue_updated_at,
		issue_key,
		issue_project,
		issue_status,
		issue_resolved_at,
		issue_priority,
		issue_summary,
		issue_description,
		issue_type,
		issue_labels,
		issue_assignee,
		issue_developer_backend,
		issue_developer_frontend,
		issue_reviewer,
		issue_product_owner,
		issue_bug_cause,
		issue_epic,
		issue_tribe,
		issue_components,
		issue_fix_versions
	)
	VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21);
	`
	rows, err := db.Query(
		query,
		s.CreatedAt,
		s.UpdatedAt,
		s.Key,
		s.Project,
		s.Status,
		s.ResolvedAt,
		s.Priority,
		s.Summary,
		s.Description,
		s.Type,
		s.Labels,
		s.Assignee,
		s.DeveloperBackend,
		s.DeveloperFrontend,
		s.Reviewer,
		s.ProductOwner,
		s.BugCause,
		s.Epic,
		s.Tribe,
		s.Components,
		s.FixVersions,
	)
	if err != nil {
		log.Fatalln(fmt.Errorf("error in `InsertIssueState`: %s", err))
	}
	rows.Close()
}

func openDB() *sql.DB {
	connStr := os.Getenv("DB_URL")
	db, err := sql.Open("postgres", connStr)
	db.SetMaxOpenConns(MaxOpenConns)
	if err != nil {
		log.Fatalln(fmt.Errorf("error in `openDB`: %s", err))
	}
	return db
}

func (db *DB) Reset() {
	queries := make([]string, 0)
	queries = append(queries, queriesResetTableJiraIssuesEvents()...)
	queries = append(queries, queriesResetTableJiraIssuesStates()...)
	err := db.doQueries(queries)
	if err != nil {
		log.Fatalln(fmt.Errorf("error in `Reset`: %s", err))
	}
}

func (db *DB) DropDBTables() {
	queries := []string{
		`DROP TABLE IF EXISTS "jira_issues_events";`,
		`DROP TABLE IF EXISTS "jira_issues_states";`,
	}
	err := db.doQueries(queries)
	if err != nil {
		log.Fatalln(fmt.Errorf("error in `dropDBTables`: %s", err))
	}
}

// doQueries performs the specified queries on the passed db.
// If an error occurs, it returns the error. This function can't
// be used for queries where you need the result rows.
func (db *DB) doQueries(queries []string) error {
	for _, q := range queries {
		rows, err := db.Query(q)
		if err != nil {
			return err
		}
		defer rows.Close()
		if err = rows.Err(); err != nil {
			return err
		}
	}
	return nil
}

func queriesResetTableJiraIssuesEvents() []string {
	return []string{
		`DROP TABLE IF EXISTS "jira_issues_events";`,
		`CREATE TABLE "jira_issues_events" (
		  "id" serial primary key not null,
		  "inserted_at" TIMESTAMP(6) NOT NULL DEFAULT statement_timestamp(),
		  "event_time" TIMESTAMP NOT NULL,
		  "event_kind" TEXT NOT NULL,
		  "event_author" TEXT NOT NULL,
	          "issue_created_at" TIMESTAMP NOT NULL,
		  "issue_updated_at" TIMESTAMP NOT NULL,
		  "issue_key" TEXT NOT NULL,
		  "issue_project" TEXT NOT NULL,
		  "issue_status" TEXT NOT NULL,
		  "issue_resolved_at" TIMESTAMP,
		  "issue_priority" TEXT NOT NULL,
		  "issue_summary" TEXT NOT NULL,
		  "issue_description" TEXT,
		  "issue_type" TEXT NOT NULL,
		  "issue_labels" TEXT,
		  "issue_assignee" TEXT,
		  "issue_developer_backend" TEXT,
		  "issue_developer_frontend" TEXT,
		  "issue_reviewer" TEXT,
		  "issue_product_owner" TEXT,
		  "issue_bug_cause" TEXT,
		  "issue_epic" TEXT,
		  "issue_tribe" TEXT,
		  "issue_components" TEXT,
		  "issue_fix_versions" TEXT,
		  "comment_body" TEXT,
		  "status_change_from" TEXT,
		  "status_change_to" TEXT
		);`,
	}
}

func queriesResetTableJiraIssuesStates() []string {
	return []string{
		`DROP TABLE IF EXISTS "jira_issues_states";`,
		`CREATE TABLE "jira_issues_states" (
			"id" SERIAL PRIMARY KEY NOT NULL,
			"inserted_at" TIMESTAMP(6) NOT NULL DEFAULT statement_timestamp(),
			"issue_created_at" TIMESTAMP NOT NULL,
			"issue_updated_at" TIMESTAMP NOT NULL,
			"issue_key" TEXT NOT NULL,
			"issue_project" TEXT NOT NULL,
			"issue_status" TEXT NOT NULL,
			"issue_resolved_at" TIMESTAMP,
			"issue_priority" TEXT NOT NULL,
			"issue_summary" TEXT NOT NULL,
			"issue_description" TEXT,
			"issue_type" TEXT NOT NULL,
			"issue_labels" TEXT,
			"issue_assignee" TEXT,
			"issue_developer_backend" TEXT,
			"issue_developer_frontend" TEXT,
			"issue_reviewer" TEXT,
			"issue_product_owner" TEXT,
			"issue_bug_cause" TEXT,
			"issue_epic" TEXT,
			"issue_tribe" TEXT,
			"issue_components" TEXT,
			"issue_fix_versions" TEXT
		);`,
	}
}
